File: tc/testprograms/TagalogDictionary.cpp
#include <vector>
#include <list>
#include <map>
#include <set>
#include <queue>
#include <deque>
#include <stack>
#include <bitset>
#include <algorithm>
#include <functional>
#include <numeric>
#include <utility>
#include <sstream>
#include <iostream>
#include <iomanip>
#include <cstdio>
#include <cmath>
#include <cstdlib>
#include <ctime>
#include <cstring>
#define all(v) (v).begin(),(v).end()
#define rall(v) (v).rbegin(),(v).rend()
#define sz size()
#define pb push_back
#define mp make_pair
#define mem(x,i) memset(x,i,sizeof(x))
#define cpresent(V,e) (find(all(V),(e))!=(V).end())
#define foreach(c,it) for(__typeof((c).begin()) it=(c).begin();it!=(c).end();it++)
#define f(i,n) for(int (i)=0;i<(int)(n);(i)++)
using namespace std;
long long toi(string s){istringstream is(s);long long x;is>>x;return x;}
string tos(long long t){stringstream st; st<<t;return st.str();}
long long gcd(long long a, long long b){return __gcd(a,b);}
long long lcm(long long a,long long b){return a*(b/gcd(a,b));}

string letters[20]={"a","b","k","d","e","g","h","i","l","m","n","ng","o","p","r","s","t","u","w","y"};
int pos(string s){
	for(int i=0;i<20;i++){
		if(letters[i]==s)
			return i;
	}
	return 0;
}
vector<int> vals(string& s){
	vector<int> vi;
	for(int i=0;i<s.sz;i++){
		if(s[i]=='n' && i+1<s.sz && s[i+1]=='g')
			vi.pb(pos("ng"));
		else
			vi.pb(pos(string(1,s[i])));
	}
	return vi;
}

bool orden(string a, string b){
	vector<int>A=vals(a);
	vector<int>B=vals(b);
	int n=min((int)a.sz,(int)b.sz);
	for(int i=0;i<n;i++){
		if(A[i]!=B[i])
			return A[i]<B[i];
	}
	return a.sz<b.sz;
}

class TagalogDictionary {
public:vector <string> sortWords(vector <string> words) {	
		sort(all(words),orden);
		return words;		
	}
	//Powered by [Ziklon]
};
// BEGIN KAWIGIEDIT TESTING
// Generated by KawigiEdit 2.1.4 (beta) modified by pivanof
bool KawigiEdit_RunTest(int testNum, vector <string> p0, bool hasAnswer, vector <string> p1) {
	cout << "Test " << testNum << ": [" << "{";
	for (int i = 0; int(p0.size()) > i; ++i) {
		if (i > 0) {
			cout << ",";
		}
		cout << "\"" << p0[i] << "\"";
	}
	cout << "}";
	cout << "]" << endl;
	TagalogDictionary *obj;
	vector <string> answer;
	obj = new TagalogDictionary();
	clock_t startTime = clock();
	answer = obj->sortWords(p0);
	clock_t endTime = clock();
	delete obj;
	bool res;
	res = true;
	cout << "Time: " << double(endTime - startTime) / CLOCKS_PER_SEC << " seconds" << endl;
	if (hasAnswer) {
		cout << "Desired answer:" << endl;
		cout << "\t" << "{";
		for (int i = 0; int(p1.size()) > i; ++i) {
			if (i > 0) {
				cout << ",";
			}
			cout << "\"" << p1[i] << "\"";
		}
		cout << "}" << endl;
	}
	cout << "Your answer:" << endl;
	cout << "\t" << "{";
	for (int i = 0; int(answer.size()) > i; ++i) {
		if (i > 0) {
			cout << ",";
		}
		cout << "\"" << answer[i] << "\"";
	}
	cout << "}" << endl;
	if (hasAnswer) {
		if (answer.size() != p1.size()) {
			res = false;
		} else {
			for (int i = 0; int(answer.size()) > i; ++i) {
				if (answer[i] != p1[i]) {
					res = false;
				}
			}
		}
	}
	if (!res) {
		cout << "DOESN'T MATCH!!!!" << endl;
	} else if (double(endTime - startTime) / CLOCKS_PER_SEC >= 2) {
		cout << "FAIL the timeout" << endl;
		res = false;
	} else if (hasAnswer) {
		cout << "Match :-)" << endl;
	} else {
		cout << "OK, but is it right?" << endl;
	}
	cout << "" << endl;
	return res;
}
int main() {
	bool all_right;
	all_right = true;
	
	vector <string> p0;
	vector <string> p1;
	
	{
	// ----- test 0 -----
	string t0[] = {"abakada","alpabet","tagalog","ako"};
			p0.assign(t0, t0 + sizeof(t0) / sizeof(t0[0]));
	string t1[] = {"abakada","ako","alpabet","tagalog"};
			p1.assign(t1, t1 + sizeof(t1) / sizeof(t1[0]));
	all_right = KawigiEdit_RunTest(0, p0, true, p1) && all_right;
	// ------------------
	}
	
	{
	// ----- test 1 -----
	string t0[] = {"ang","ano","anim","alak","alam","alab"};
			p0.assign(t0, t0 + sizeof(t0) / sizeof(t0[0]));
	string t1[] = {"alab","alak","alam","anim","ano","ang"};
			p1.assign(t1, t1 + sizeof(t1) / sizeof(t1[0]));
	all_right = KawigiEdit_RunTest(1, p0, true, p1) && all_right;
	// ------------------
	}
	
	{
	// ----- test 2 -----
	string t0[] = {"siya","niya","kaniya","ikaw","ito","iyon"};
			p0.assign(t0, t0 + sizeof(t0) / sizeof(t0[0]));
	string t1[] = {"kaniya","ikaw","ito","iyon","niya","siya"};
			p1.assign(t1, t1 + sizeof(t1) / sizeof(t1[0]));
	all_right = KawigiEdit_RunTest(2, p0, true, p1) && all_right;
	// ------------------
	}
	
	{
	// ----- test 3 -----
	string t0[] = {"kaba","baka","naba","ngipin","nipin"};
			p0.assign(t0, t0 + sizeof(t0) / sizeof(t0[0]));
	string t1[] = {"baka","kaba","naba","nipin","ngipin"};
			p1.assign(t1, t1 + sizeof(t1) / sizeof(t1[0]));
	all_right = KawigiEdit_RunTest(3, p0, true, p1) && all_right;
	// ------------------
	}
	
	{
	// ----- test 4 -----
	string t0[] = {"knilngiggnngginggn","ingkigningg","kingkong","dingdong","dindong","dingdont","ingkblot"};
			p0.assign(t0, t0 + sizeof(t0) / sizeof(t0[0]));
	string t1[] = {"kingkong","knilngiggnngginggn","dindong","dingdont","dingdong","ingkblot","ingkigningg"};
			p1.assign(t1, t1 + sizeof(t1) / sizeof(t1[0]));
	all_right = KawigiEdit_RunTest(4, p0, true, p1) && all_right;
	// ------------------
	}
	
	{
	// ----- test 5 -----
	string t0[] = {"silangang","baka","bada","silang"};
			p0.assign(t0, t0 + sizeof(t0) / sizeof(t0[0]));
	string t1[] = {"baka","bada","silang","silangang"};
			p1.assign(t1, t1 + sizeof(t1) / sizeof(t1[0]));
	all_right = KawigiEdit_RunTest(5, p0, true, p1) && all_right;
	// ------------------
	}
	
	if (all_right) {
		cout << "You're a stud (at least on the example cases)!" << endl;
	} else {
		cout << "Some of the test cases had errors." << endl;
	}
	return 0;
}
// END KAWIGIEDIT TESTING
//Powered by KawigiEdit 2.1.4 (beta) modified by pivanof!

File: tc/testprograms/TagalogDictionary.java
// BEGIN CUT HERE
// PROBLEM STATEMENT
// 
// In the first half of the 20th century, around the time that Tagalog became the national 
// language of the Philippines, a standardized alphabet was introduced to be used in 
// Tagalog school books (since then, the national language has changed to a hybrid 
// "Pilipino" language, as Tagalog is only one of several major languages spoken in the 
// Philippines).
// 
// 
// Tagalog's 20-letter alphabet is as follows:
// 
// a b k d e g h i l m n ng o p r s t u w y
// 
// Note that not all letters used in English are present, 'k' is the third letter, and 'ng' 
// is a single letter that comes between 'n' and 'o'.
// 
// 
// You are compiling a Tagalog dictionary, and for people to be able to find words in it, 
// the words need to be sorted in alphabetical order with reference to the Tagalog 
// alphabet.  Given a list of Tagalog words as a String[], return the same list in Tagalog 
// alphabetical order.
// 
// 
// DEFINITION
// Class:TagalogDictionary
// Method:sortWords
// Parameters:String[]
// Returns:String[]
// Method signature:String[] sortWords(String[] words)
// 
// 
// NOTES
// -Any 'n' followed followed by a 'g' should be considered a single 'ng' letter (the one 
// that comes between 'n' and 'o').
// 
// 
// CONSTRAINTS
// -words will contain between 1 and 50 elements, inclusive.
// -Each element of words will contain between 1 and 50 characters, inclusive.
// -Each character of each element of words will be a valid lowercase letter that appears 
// in the Tagalog alphabet.
// -Each element of words will be distinct.
// 
// 
// EXAMPLES
// 
// 0)
// {"abakada","alpabet","tagalog","ako"}
// 
// Returns: {"abakada", "ako", "alpabet", "tagalog" }
// 
// The tagalog word for "alphabet", a Tagalogization of the English word "alphabet", the 
// name of the language, and the word for "I".
// 
// 1)
// {"ang","ano","anim","alak","alam","alab"}
// 
// Returns: {"alab", "alak", "alam", "anim", "ano", "ang" }
// 
// A few "A" words that are alphabetically close together.
// 
// 2)
// {"siya","niya","kaniya","ikaw","ito","iyon"}
// 
// Returns: {"kaniya", "ikaw", "ito", "iyon", "niya", "siya" }
// 
// Common Tagalog pronouns.
// 
// 3)
// {"kaba","baka","naba","ngipin","nipin"}
// 
// Returns: {"baka", "kaba", "naba", "nipin", "ngipin" }
// 
// 4)
// {"knilngiggnngginggn","ingkigningg","kingkong","dingdong","dindong","dingdont","ingkblot"}
// 
// Returns: {"kingkong", "knilngiggnngginggn", "dindong", "dingdont", "dingdong", 
// "ingkblot", "ingkigningg" }
// 
// 5)
// {"silangang", "baka", "bada", "silang"}
// 
// Returns: {"baka", "bada", "silang", "silangang" }
// 
// 
// 
// END CUT HERE
import java.util.*;
import java.util.regex.*;
import java.text.*;
import java.math.*;
import java.awt.geom.*;
import static java.lang.Math.*;
import static java.util.Arrays.*;
import static java.lang.Integer.*;
import static java.lang.Long.*;
import static java.lang.Character.*;
import static java.lang.Double.*;
import static java.util.Collections.*;

public class TagalogDictionary {
	
	String letters="abkdeghilmncoprstuwy";
	public String[] sortWords(String[] words) {
		for(int i=0;i<words.length;i++)
			words[i]=words[i].replaceAll("ng", "c");
		
		Arrays.sort(words,new myCompare());		
		for(int i=0;i<words.length;i++)
			words[i]=words[i].replaceAll("c", "ng");
		return words;
	}
	
	class myCompare implements Comparator<String>{

		
		public int compare(String a, String b) {
			if(a.equals(b))return 0;
			int m=min(a.length(),b.length());
			for(int i=0;i<m;i++){
				int p=letters.indexOf(a.charAt(i));
				int q=letters.indexOf(b.charAt(i));
				if(p<q)return -1;
				if(p>q)return 1;
				
			}
			if(a.length()<b.length())return -1;
			return 1;
		}
		
		
	}
	
	// BEGIN CUT HERE
	public void print(String s) {
		System.err.print("   " + s);
	}
	
	public void println(String s) {
		print(s + "\n");
	}
	
	public static void main(String[] args) {
		if (args.length == 0) {
			TagalogDictionaryHarness.run_test(-1);
		} else {
			for (int i = 0; i < args.length; ++i)
				TagalogDictionaryHarness.run_test(Integer.valueOf(args[i]));
		}
	}
	// END CUT HERE
}

// BEGIN CUT HERE
class TagalogDictionaryHarness {
	public static void run_test(int casenum) {
		if (casenum != -1) {
			if (runTestCase(casenum) == -1)
				System.err.println("Illegal input! Test case " + casenum
						+ " does not exist.");
			return;
		}
		
		int correct = 0, total = 0;
		for (int i = 0;; ++i) {
			int x = runTestCase(i);
			if (x == -1) {
				if (i >= 100) break;
				continue;
			}
			correct += x;
			++total;
		}
		
		if (total == 0) {
			System.err.println("No test cases run.");
		} else if (correct < total) {
			System.err.println("Some cases FAILED (passed " + correct + " of "
					+ total + ").");
		} else {
			System.err.println("All " + total + " tests passed!");
		}
	}
	
	static boolean compareOutput(String[] expected, String[] result) {
		if (expected.length != result.length) return false;
		for (int i = 0; i < expected.length; ++i)
			if (!expected[i].equals(result[i])) return false;
		return true;
	}
	
	static String formatResult(String[] res) {
		String ret = "";
		ret += "{";
		for (int i = 0; i < res.length; ++i) {
			if (i > 0) ret += ",";
			ret += String.format(" \"%s\"", res[i]);
		}
		ret += " }";
		return ret;
	}
	
	static int verifyCase(int casenum, String[] expected, String[] received) {
		System.err.print("Example " + casenum + "... ");
		if (compareOutput(expected, received)) {
			System.err.println("PASSED");
			return 1;
		} else {
			System.err.println("FAILED");
			System.err.println("    Expected: " + formatResult(expected));
			System.err.println("    Received: " + formatResult(received));
			return 0;
		}
	}
	
	static int runTestCase(int casenum) {
		switch (casenum) {
			case 0: {
				String[] words = { "abakada", "alpabet", "tagalog", "ako" };
				String[] expected__ = { "abakada", "ako", "alpabet", "tagalog" };
				
				return verifyCase(casenum, expected__,
						new TagalogDictionary().sortWords(words));
			}
			case 1: {
				String[] words = { "ang", "ano", "anim", "alak", "alam", "alab" };
				String[] expected__ = { "alab", "alak", "alam", "anim", "ano",
						"ang" };
				
				return verifyCase(casenum, expected__,
						new TagalogDictionary().sortWords(words));
			}
			case 2: {
				String[] words = { "siya", "niya", "kaniya", "ikaw", "ito",
						"iyon" };
				String[] expected__ = { "kaniya", "ikaw", "ito", "iyon",
						"niya", "siya" };
				
				return verifyCase(casenum, expected__,
						new TagalogDictionary().sortWords(words));
			}
			case 3: {
				String[] words = { "kaba", "baka", "naba", "ngipin", "nipin" };
				String[] expected__ = { "baka", "kaba", "naba", "nipin",
						"ngipin" };
				
				return verifyCase(casenum, expected__,
						new TagalogDictionary().sortWords(words));
			}
			case 4: {
				String[] words = { "knilngiggnngginggn", "ingkigningg",
						"kingkong", "dingdong", "dindong", "dingdont",
						"ingkblot" };
				String[] expected__ = { "kingkong", "knilngiggnngginggn",
						"dindong", "dingdont", "dingdong", "ingkblot",
						"ingkigningg" };
				
				return verifyCase(casenum, expected__,
						new TagalogDictionary().sortWords(words));
			}
			case 5: {
				String[] words = { "silangang", "baka", "bada", "silang" };
				String[] expected__ = { "baka", "bada", "silang", "silangang" };
				
				return verifyCase(casenum, expected__,
						new TagalogDictionary().sortWords(words));
			}
			
			// custom cases
			
			/*
			 * case 6: { String[] words = ; String[] expected__ = ;
			 * 
			 * return verifyCase(casenum, expected__, new
			 * TagalogDictionary().sortWords(words)); }
			 */
			/*
			 * case 7: { String[] words = ; String[] expected__ = ;
			 * 
			 * return verifyCase(casenum, expected__, new
			 * TagalogDictionary().sortWords(words)); }
			 */
			/*
			 * case 8: { String[] words = ; String[] expected__ = ;
			 * 
			 * return verifyCase(casenum, expected__, new
			 * TagalogDictionary().sortWords(words)); }
			 */
			default:
				return -1;
		}
	}
}

// END CUT HERE
